pysr/regressor_sequence.py

```diff
@@ -0,0 +1,381 @@
+import copy
+import os
+import pickle as pkl
+import re
+import shutil
+import sys
+import tempfile
+import warnings
+
+from typing import Any, Callable, Dict, List, Literal, Optional, Tuple, Union, cast
+
+import numpy as np
+from numpy import ndarray
+from numpy.typing import NDArray
+from sklearn.base import BaseEstimator
+from pathlib import Path
+
+from .sr import PySRRegressor
+from .utils import (
+    ArrayLike,
+    PathLike,
+    _csv_filename_to_pkl_filename,
+    _preprocess_julia_floats,
+    _safe_check_feature_names_in,
+    _subscriptify,
+    _suggest_keywords,
+)
+
+
+def _check_assertions(
+    X,
+    y=None,
+    Xresampled=None,
+    recursive_history_length=None,
+    weights=None,
+    variable_names=None,
+    X_units=None,
+    y_units=None,
+):
+    if recursive_history_length is not None and recursive_history_length <= 0:
+        raise ValueError(
+            "The `recursive_history_length` parameter must be greater than 0 (otherwise it's not recursion)."
+        )
+    if len(X.shape) > 2:
+        raise ValueError(
+            "Recursive symbolic regression only supports up to 2D data; please flatten your data first"
+        )
+    elif len(X) < 2:
+        raise ValueError(
+            "Recursive symbolic regression requires at least 2 datapoints; if you tried to pass a 1D array, use array.reshape(-1, 1)"
+        )
+    if len(X) <= recursive_history_length + 1:
+        raise ValueError(
+            f"Recursive symbolic regression with a history length of {recursive_history_length} requires at least {recursive_history_length + 2} datapoints."
+        )
+    if isinstance(weights, np.ndarray) and len(weights) != len(X):
+        raise ValueError("The length of `weights` must have shape (n_times,).")
+    if isinstance(variable_names, list) and len(variable_names) != X.shape[1]:
+        raise ValueError(
+            "The length of `variable_names` must be equal to the number of features in `X`."
+        )
+    if isinstance(X_units, list) and len(X_units) != X.shape[1]:
+        raise ValueError(
+            "The length of `X_units` must be equal to the number of features in `X`."
+        )
+    if y is not None:
+        raise ValueError("Recursive symbolic regression does not use `y`")
+    if y_units is not None:
+        raise ValueError("Recursive symbolic regression does not use `y_units`")
+    if Xresampled is not None:
+        raise ValueError("Recursive symbolic regression does not use `Xresampled`")
+
+
+class PySRSequenceRegressor(BaseEstimator):
```

Comment:
Are there other sklearn base classes to use here? Or do we only want `BaseEstimator`?

Reply:
well PySRRegressor inherits from MultiOutputMixin and RegressorMixin so we could use those

Reply:
I'm not sure it fits those though, maybe read about the requirements for those abstract classes? i.e., the scikit-learn docs.

Reply:
I think it does fit RegressorMixin

> Mixin class for all regression estimators in scikit-learn.

Reply:
I found one line of documentation
PySRSequenceRegressor does require multi output for multidimensional recurrent sequences, right?
> Mixin to mark estimators that support multioutput.